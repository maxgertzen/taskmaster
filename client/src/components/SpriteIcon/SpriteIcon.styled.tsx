import styled from '@emotion/styled';

export const StyledIconWrapper = styled.div<{
  size: number;
  isVisible: boolean;
  applyInvert?: boolean;
  isClickable?: boolean;
}>(
  ({
    theme,
    size,
    applyInvert = false,
    isVisible = true,
    isClickable = true,
  }) => ({
    display: 'flex',
    width: theme.spacing(size),
    height: theme.spacing(size),
    justifyContent: 'center',
    alignItems: 'center',
    cursor: isClickable ? 'pointer' : 'default',
    opacity: isVisible ? 1 : 0.5,
    transition: 'filter 0.15s ease-in',
    svg: {
      filter: `contrast(1.5) ${applyInvert ? 'invert(1)' : ''}`,
    },
    ':hover': {
      filter: 'hue-rotate(90deg) contrast(1.8)',
    },
  })
);
