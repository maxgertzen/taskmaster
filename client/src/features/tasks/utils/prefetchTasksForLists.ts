import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
