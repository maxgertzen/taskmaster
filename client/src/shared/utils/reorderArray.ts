export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
