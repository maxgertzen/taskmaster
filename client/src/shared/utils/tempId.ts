export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
