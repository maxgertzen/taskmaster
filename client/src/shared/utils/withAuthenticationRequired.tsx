import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
