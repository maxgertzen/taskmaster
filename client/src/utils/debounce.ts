export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
