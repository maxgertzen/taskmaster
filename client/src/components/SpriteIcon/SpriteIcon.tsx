/// <reference types="vite-plugin-svgr/client" />
import { FC, FunctionComponent, useMemo } from 'react';

import a2z from '../../assets/sprites/icons/a2z.svg?react';
import closepanel from '../../assets/sprites/icons/closepanel.svg?react';
import dragDark from '../../assets/sprites/icons/drag-dm.svg?react';
import drag from '../../assets/sprites/icons/drag.svg?react';
import hamburgerMenuDark from '../../assets/sprites/icons/hamburger-menu-dm.svg?react';
import hamburgerMenu from '../../assets/sprites/icons/hamburger-menu.svg?react';
import listDark from '../../assets/sprites/icons/list-dm.svg?react';
import list from '../../assets/sprites/icons/list.svg?react';
import logout from '../../assets/sprites/icons/logout.svg?react';
import magnifyingDark from '../../assets/sprites/icons/magnifying-dm.svg?react';
import magnifying from '../../assets/sprites/icons/magnifying.svg?react';
import moon from '../../assets/sprites/icons/moon.svg?react';
import openpanel from '../../assets/sprites/icons/openpanel.svg?react';
import pencilDark from '../../assets/sprites/icons/pencil-dm.svg?react';
import pencil from '../../assets/sprites/icons/pencil.svg?react';
import plusDark from '../../assets/sprites/icons/plus-dm.svg?react';
import plus from '../../assets/sprites/icons/plus.svg?react';
import sun from '../../assets/sprites/icons/sun.svg?react';
import threeDotsDark from '../../assets/sprites/icons/three-dots-dm.svg?react';
import threeDots from '../../assets/sprites/icons/three-dots.svg?react';
import trashDark from '../../assets/sprites/icons/trash-dm.svg?react';
import trash from '../../assets/sprites/icons/trash.svg?react';
import userDark from '../../assets/sprites/icons/user-dm.svg?react';
import user from '../../assets/sprites/icons/user.svg?react';
import xButtonDark from '../../assets/sprites/icons/x-button-dm.svg?react';
import xButton from '../../assets/sprites/icons/x-button.svg?react';
import z2a from '../../assets/sprites/icons/z2a.svg?react';
import { useThemeStore } from '../../store/themeStore';
import { IconName } from '../../types/icons';

import { StyledIconWrapper } from './SpriteIcon.styled';

interface SpriteIconProps {
  name: IconName;
  alt?: string;
  size?: number;
  isVisible?: boolean;
  onClick?: () => void;
}

const iconMap: Record<
  IconName,
  FunctionComponent<React.SVGProps<SVGSVGElement>>
> = {
  a2z,
  closepanel,
  drag,
  dragDark,
  'hamburger-menu': hamburgerMenu,
  'hamburger-menuDark': hamburgerMenuDark,
  list,
  listDark,
  logout,
  magnifying,
  magnifyingDark,
  moon,
  openpanel,
  pencil,
  pencilDark,
  plus,
  plusDark,
  sun,
  'three-dots': threeDots,
  'three-dotsDark': threeDotsDark,
  trash,
  trashDark,
  user,
  userDark,
  'x-button': xButton,
  'x-buttonDark': xButtonDark,
  z2a,
};

export const SpriteIcon: FC<SpriteIconProps> = ({
  name,
  alt = 'icon',
  size = 3,
  isVisible = true,
  onClick,
}) => {
  const mode = useThemeStore((state) => state.theme);

  const Svg = useMemo(() => {
    if (mode === 'dark' && iconMap[`${name}Dark` as IconName]) {
      return iconMap[`${name}Dark` as IconName];
    }
    return iconMap[name];
  }, [mode, name]);

  const applyInvert = useMemo(() => {
    return (
      mode === 'dark' && name !== 'sun' && !iconMap[`${name}Dark` as IconName]
    );
  }, [mode, name]);

  return (
    <StyledIconWrapper
      applyInvert={applyInvert}
      size={size}
      role='button'
      isVisible={isVisible}
    >
      <Svg height='100%' style={{ position: 'relative' }} onClick={onClick}>
        <title>{alt}</title>
      </Svg>
    </StyledIconWrapper>
  );
};
