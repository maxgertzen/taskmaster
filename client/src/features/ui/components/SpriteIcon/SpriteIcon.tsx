import { FC, MouseEventHandler, useMemo } from 'react';

import { iconMap } from '@/shared/assets/icons';
import { useThemeStore } from '@/shared/store/themeStore';
import { IconName } from '@/shared/types/icons';

import { StyledIconWrapper } from './SpriteIcon.styled';

interface SpriteIconProps {
  name: IconName;
  alt?: string;
  size?: number;
  isVisible?: boolean;
  onClick?: MouseEventHandler<HTMLDivElement>;
  'data-testid'?: string;
}

export const SpriteIcon: FC<SpriteIconProps> = ({
  name,
  alt = 'icon',
  size = 3,
  isVisible = true,
  'data-testid': dataTestId,
  onClick,
}) => {
  const mode = useThemeStore((state) => state.theme);

  const Svg = useMemo(() => {
    if (mode === 'dark' && iconMap[`${name}Dark` as IconName]) {
      return iconMap[`${name}Dark` as IconName];
    }
    return iconMap[name];
  }, [mode, name]);

  const applyInvert = useMemo(() => {
    return (
      mode === 'dark' && name !== 'sun' && !iconMap[`${name}Dark` as IconName]
    );
  }, [mode, name]);

  const handleOnClick = (e: React.MouseEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    onClick?.(e);
  };

  return (
    <StyledIconWrapper
      applyInvert={applyInvert}
      data-testid={dataTestId ?? `icon-${name}`}
      size={size}
      role='button'
      isVisible={isVisible}
      onClick={onClick ? handleOnClick : undefined}
    >
      <Svg height='100%' style={{ position: 'relative' }}>
        <title>{alt}</title>
      </Svg>
    </StyledIconWrapper>
  );
};
