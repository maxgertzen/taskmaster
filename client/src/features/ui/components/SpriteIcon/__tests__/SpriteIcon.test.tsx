import { fireEvent } from '@testing-library/dom';
import { describe, it, expect, vi, beforeEach } from 'vitest';

import {
  mockThemeStore,
  mockThemeStoreMode,
} from '@/tests/__mocks__/storeMocks';
import { forceClick, render, screen } from '@/tests/utils';

import { SpriteIcon } from '../SpriteIcon';

vi.mock('@/shared/store/themeStore', () => ({
  useThemeStore: mockThemeStore,
}));

describe('SpriteIcon', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe.each([['light'], ['dark']])('in %s mode', (theme) => {
    beforeEach(() => {
      mockThemeStoreMode.mockReturnValue(theme);
    });

    it('renders basic icon correctly', () => {
      const { container } = render(<SpriteIcon name='plus' alt='add item' />);
      expect(container).toMatchSnapshot();
    });

    it('renders with custom size', () => {
      const { container } = render(<SpriteIcon name='plus' size={5} />);
      expect(container).toMatchSnapshot();
    });

    it('handles click events', () => {
      const mockOnClick = vi.fn();
      render(<SpriteIcon name='plus' onClick={mockOnClick} />);

      const icon = screen.getByTestId('icon-plus');
      fireEvent.click(icon);

      expect(mockOnClick).toHaveBeenCalled();
    });

    it('prevents default and stops propagation on click', async () => {
      const mockOnClick = vi.fn();

      render(<SpriteIcon name='plus' onClick={mockOnClick} />);

      const icon = screen.getByTestId('icon-plus');
      const { preventDefaultMock, stopPropagationMock } = forceClick(icon);

      expect(preventDefaultMock).toHaveBeenCalled();
      expect(stopPropagationMock).toHaveBeenCalled();
    });

    it('handles visibility prop', () => {
      const { container } = render(
        <SpriteIcon name='plus' isVisible={false} />
      );
      expect(container).toMatchSnapshot();
    });

    it('uses dark variant of icon when available in dark mode', () => {
      const { container } = render(<SpriteIcon name='plus' />);
      expect(container).toMatchSnapshot();
    });
  });
});
