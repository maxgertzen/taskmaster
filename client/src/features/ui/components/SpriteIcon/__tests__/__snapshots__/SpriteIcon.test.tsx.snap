// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SpriteIcon > in dark mode > handles visibility prop 1`] = `
<div>
  <div
    class="css-j6qdau"
    data-testid="icon-plus"
    role="button"
    size="3"
  >
    <svg
      data-testid="plusDark-icon"
    />
  </div>
</div>
`;

exports[`SpriteIcon > in dark mode > renders basic icon correctly 1`] = `
<div>
  <div
    class="css-4dcg29"
    data-testid="icon-plus"
    role="button"
    size="3"
  >
    <svg
      data-testid="plusDark-icon"
    />
  </div>
</div>
`;

exports[`SpriteIcon > in dark mode > renders with custom size 1`] = `
<div>
  <div
    class="css-1tr2yd6"
    data-testid="icon-plus"
    role="button"
    size="5"
  >
    <svg
      data-testid="plusDark-icon"
    />
  </div>
</div>
`;

exports[`SpriteIcon > in dark mode > uses dark variant of icon when available in dark mode 1`] = `
<div>
  <div
    class="css-4dcg29"
    data-testid="icon-plus"
    role="button"
    size="3"
  >
    <svg
      data-testid="plusDark-icon"
    />
  </div>
</div>
`;

exports[`SpriteIcon > in light mode > handles visibility prop 1`] = `
<div>
  <div
    class="css-j6qdau"
    data-testid="icon-plus"
    role="button"
    size="3"
  >
    <svg
      data-testid="plus-icon"
    />
  </div>
</div>
`;

exports[`SpriteIcon > in light mode > renders basic icon correctly 1`] = `
<div>
  <div
    class="css-4dcg29"
    data-testid="icon-plus"
    role="button"
    size="3"
  >
    <svg
      data-testid="plus-icon"
    />
  </div>
</div>
`;

exports[`SpriteIcon > in light mode > renders with custom size 1`] = `
<div>
  <div
    class="css-1tr2yd6"
    data-testid="icon-plus"
    role="button"
    size="5"
  >
    <svg
      data-testid="plus-icon"
    />
  </div>
</div>
`;

exports[`SpriteIcon > in light mode > uses dark variant of icon when available in dark mode 1`] = `
<div>
  <div
    class="css-4dcg29"
    data-testid="icon-plus"
    role="button"
    size="3"
  >
    <svg
      data-testid="plus-icon"
    />
  </div>
</div>
`;
